File: kotlin/Algorithm/src/main/kotlin/Boj5430.kt
fun main() {
    repeat(readln().toInt()) {
        val commands = readln()
        val dataLength = readln().toInt()
        val inputData: String = readln()

        if (dataLength == 0 && commands.contains('D')) {
            println("error")
            return@repeat
        }

        var isReversed = false
        var startIdx = 0
        var endIdx: Int = dataLength - 1

        val data = if (dataLength > 0) inputData.removeSurrounding("[", "]")
            .split(',')
            .map { it.toInt() }
            .toTypedArray() else emptyArray()

        for (command in commands) {
            when (command) {
                'R' -> isReversed = !isReversed
                'D' -> {
                    if (startIdx > endIdx) {
                        print("error")
                        return
                    }
                    if (isReversed) endIdx-- else startIdx++
                }
            }
        }

        printArray(data, startIdx, endIdx, isReversed)
    }
}

fun printArray(array: Array<Int>, startIdx: Int, endIdx: Int, isReversed: Boolean) {
    val result = if (isReversed) {
        array.sliceArray(startIdx..endIdx).reversed().joinToString(",")
    } else {
        array.sliceArray(startIdx..endIdx).joinToString(",")
    }

    println("[$result]")
}

File: dart/algorithm/bin/boj_5430.dart
import 'dart:io';

void main(List<String> arguments) {
  int T = int.parse(stdin.readLineSync()!);

  for (int i = 0; i < T; i++) {
    String commands = stdin.readLineSync()!;
    int dataLength = int.parse(stdin.readLineSync()!);

    bool isError = false;
    bool isReversed = false;
    int startIdx = 0;
    int endIdx = dataLength - 1;

    String inputData = stdin.readLineSync()!;

    if (dataLength == 0) {
      if (commands.contains('D')) {
        print('error');
        return;
      }
    }

    List<int> data = inputData
        .replaceAll(RegExp(r'[\[\]]'), '')
        .split(',')
        .map((str) => int.parse(str))
        .toList();

    for (var command in commands.runes) {
      if (command == 'R'.runes.single) {
        isReversed = !isReversed;
      } else if (command == 'D'.runes.single) {
        if (startIdx > endIdx) {
          isError = true;
          break;
        } else {
          if (isReversed) {
            endIdx -= 1;
          } else {
            startIdx += 1;
          }
        }
      }
    }

    if (isError) {
      stdout.write("error\n");
    } else if (startIdx > endIdx) {
      stdout.write("[]\n");
    } else if (isReversed) {
      stdout.write("[");
      for (int idx = endIdx; idx > startIdx; idx--) {
        stdout.write("${data[idx]},");
      }
      stdout.write("${data[startIdx]}]\n");
    } else {
      stdout.write("[");
      for (int idx = startIdx; idx < endIdx; idx++) {
        stdout.write("${data[idx]},");
      }
      stdout.write("${data[endIdx]}]\n");
    }
  }
}
